kleingebaeck.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "KLEINGEBAECK 1"
.TH KLEINGEBAECK 1 "2025-02-27" "1" "User Commands"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
kleingebaeck \- kleinanzeigen.de backup tool
.SH "SYNOPSYS"
.IX Header "SYNOPSYS"
.Vb 10
\&    Usage: kleingebaeck [\-dvVhmoc] [<ad\-listing\-url>,...]
\&    Options:
\&    \-u \-\-user    <uid>      Backup ads from user with uid <uid>.
\&    \-d \-\-debug              Enable debug output.
\&    \-v \-\-verbose            Enable verbose output.
\&    \-o \-\-outdir  <dir>      Set output dir (default: current directory)
\&    \-l \-\-limit   <num>      Limit the ads to download to <num>, default: load all.
\&    \-c \-\-config  <file>     Use config file <file> (default: ~/.kleingebaeck).
\&       \-\-ignoreerrors       Ignore HTTP errors, may lead to incomplete ad backup.
\&    \-f \-\-force              Overwrite images and ads even if the already exist.
\&    \-m \-\-manual             Show manual.
\&    \-h \-\-help               Show usage.
\&    \-V \-\-version            Show program version.
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This tool can be used to backup ads on the german ad page <https://kleinanzeigen.de>.
.PP
It downloads all (or  only the specified ones) ads of  one user into a
directory, each ad into its own subdirectory. The backup will contain
a textfile \fBAdlisting.txt\fR which contains the ad contents such as
title, body, price etc. All images will be downloaded as well.
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
You   can  create   a  config   file  to   save  typing.   By  default
\&\f(CW\*(C`~/.kleingebaeck\*(C'\fR is being used but you can specify one with \f(CW\*(C`\-c\*(C'\fR as
well.   We    use   \s-1TOML\s0   as   our    configuration   language.   See
<https://toml.io/en/>.
.PP
Format is pretty simple:
.PP
.Vb 11
\&    user = 1010101
\&    loglevel = verbose
\&    outdir = "test"
\&    useragent = "Mozilla/5.0"
\&    template = """
\&    Title: {{.Title}}
\&    Price: {{.Price}}
\&    Id: {{.ID}}
\&    Category: {{.Category}}
\&    Condition: {{.Condition}}
\&    Created: {{.Created}}
\&
\&    {{.Text}}
\&    """
.Ve
.PP
Be careful  if you want  to change the  template.  The variable  is a
multiline string surrounded  by three double quotes. You  can left out
certain   fields  and   use  any   formatting  you   like.  Refer   to
<https://pkg.go.dev/text/template>  for   details  how  to   write  a
template. Also read the \s-1TEMPLATES\s0 section below.
.PP
If you're on  windows and want to customize the  output directory, put
it into single  quotes to avoid the backslashes  interpreted as escape
chars like this:
.PP
.Vb 1
\&    outdir = \*(AqC:\eData\eAds\*(Aq
.Ve
.SH "TEMPLATES"
.IX Header "TEMPLATES"
Various parts  of the configuration  can be modified  using templates:
the output directory, the ad directory and the ad listing itself.
.SS "\s-1OUTPUT DIR TEMPLATE\s0"
.IX Subsection "OUTPUT DIR TEMPLATE"
The config varialbe \f(CW\*(C`outdir\*(C'\fR or the command line parameter \f(CW\*(C`\-o\*(C'\fR take a
template which may contain:
.ie n .IP """{{.Year}}""" 4
.el .IP "\f(CW{{.Year}}\fR" 4
.IX Item "{{.Year}}"
.PD 0
.ie n .IP """{{.Month}}""" 4
.el .IP "\f(CW{{.Month}}\fR" 4
.IX Item "{{.Month}}"
.ie n .IP """{{.Day}}""" 4
.el .IP "\f(CW{{.Day}}\fR" 4
.IX Item "{{.Day}}"
.PD
.PP
That way you can create a new output directory for every backup
run. For example:
.PP
.Vb 1
\&    outdir = "/home/backups/ads\-{{.Year}}\-{{.Month}}\-{{.Day}}"
.Ve
.PP
Or using the command line flag:
.PP
.Vb 1
\&    \-o "/home/backups/ads\-{{.Year}}\-{{.Month}}\-{{.Day}}"
.Ve
.PP
The default value is \f(CW\*(C`.\*(C'\fR \- the current directory.
.SS "\s-1AD DIRECTORY TEMPLATE\s0"
.IX Subsection "AD DIRECTORY TEMPLATE"
The ad directory name can be modified using the following ad values:
.IP "{{.Price}}" 4
.IX Item "{{.Price}}"
.PD 0
.IP "{{.ID}}" 4
.IX Item "{{.ID}}"
.IP "{{.Category}}" 4
.IX Item "{{.Category}}"
.IP "{{.Condition}}" 4
.IX Item "{{.Condition}}"
.IP "{{.Created}}" 4
.IX Item "{{.Created}}"
.IP "{{.Slug}}" 4
.IX Item "{{.Slug}}"
.IP "{{.Text}}" 4
.IX Item "{{.Text}}"
.PD
.PP
It can only be configured in the config file. By default only
\&\f(CW\*(C`{{.Slug}}\*(C'\fR is being used, this is the title of the ad in url format.
.SS "\s-1AD NAME TEMPLATE\s0"
.IX Subsection "AD NAME TEMPLATE"
The name of the directory per ad can be tuned as well:
.ie n .IP """{{.Year}}""" 4
.el .IP "\f(CW{{.Year}}\fR" 4
.IX Item "{{.Year}}"
.PD 0
.ie n .IP """{{.Month}}""" 4
.el .IP "\f(CW{{.Month}}\fR" 4
.IX Item "{{.Month}}"
.ie n .IP """{{.Day}}""" 4
.el .IP "\f(CW{{.Day}}\fR" 4
.IX Item "{{.Day}}"
.ie n .IP """{{.Slug}}""" 4
.el .IP "\f(CW{{.Slug}}\fR" 4
.IX Item "{{.Slug}}"
.ie n .IP """{{.Category}}""" 4
.el .IP "\f(CW{{.Category}}\fR" 4
.IX Item "{{.Category}}"
.ie n .IP """{{.ID}}""" 4
.el .IP "\f(CW{{.ID}}\fR" 4
.IX Item "{{.ID}}"
.PD
.SS "\s-1AD TEMPLATE\s0"
.IX Subsection "AD TEMPLATE"
The ad listing itself can be modified as well, using the same
variables as the ad name template above.
.PP
This is the default template:
.PP
.Vb 8
\&    Title: {{.Title}}
\&    Price: {{.Price}}
\&    Id: {{.ID}}
\&    Category: {{.Category}}
\&    Condition: {{.Condition}}
\&    Type: {{.Type}}
\&    Created: {{.Created}}
\&    Expire: {{.Expire}}
\&    
\&    {{.Text}}
.Ve
.PP
The config parameter to modify is \f(CW\*(C`template\*(C'\fR. See example.conf in the
source repository. Please take care, since this is a multiline
string. This is how it shall look if you modify it:
.PP
.Vb 2
\&    template="""
\&    Title: {{.Title}}
\&    
\&    {{.Text}}
\&    """
.Ve
.PP
That is, the content between the two \f(CW"""\fR chars is the template.
.SH "SETUP"
.IX Header "SETUP"
To setup the tool, you need to lookup your userid on
kleinanzeigen.de. Go to your ad overview page while \s-1NOT\s0 being logged
in:
.PP
.Vb 1
\&    https://www.kleinanzeigen.de/s\-bestandsliste.html?userId=XXXXXX
.Ve
.PP
The \fB\s-1XXXXX\s0\fR part is your userid.
.PP
Put it into the configfile as outlined above. Also specify an output
directory. Then just execute \f(CW\*(C`kleingebaeck\*(C'\fR.
.PP
You can use the \fB\-v\fR option to  get verbose output or \fB\-d\fR to enable
debugging.
.SH "ENVIRONMENT VARIABLES"
.IX Header "ENVIRONMENT VARIABLES"
The following environment variables  are considered:
.PP
.Vb 7
\&    KLEINGEBAECK_USER
\&    KLEINGEBAECK_DEBUG
\&    KLEINGEBAECK_VERBOSE
\&    KLEINGEBAECK_OUTDIR
\&    KLEINGEBAECK_LIMIT
\&    KLEINGEBAECK_CONFIG
\&    KLEINGEBAECK_IGNOREERRORS
.Ve
.PP
Please  note,  that  they  take   precedence  over  config  file,  but
commandline flags take precedence over env!
.SH "BUGS"
.IX Header "BUGS"
In order to report a bug, unexpected behavior, feature requests
or to submit a patch, please open an issue on github:
<https://codeberg.org/scip/kleingebaeck/issues>.
.PP
Please repeat the failing command with debugging enabled \f(CW\*(C`\-d\*(C'\fR and
include the output in the issue.
.SH "LIMITATIONS"
.IX Header "LIMITATIONS"
The \f(CW\*(C`kleingebaeck\*(C'\fR doesn't currently check if it has downloaded a
file already, so it downloads everything again every time you execute
it. Be aware of it. This will change in the future.
.PP
Also there's currently no parallelization implemented. This will
change in the future.
.SH "LICENSE"
.IX Header "LICENSE"
Copyright 2023\-2025 Thomas von Dein
.PP
This program is free software: you can redistribute it and/or modify
it under the terms of the \s-1GNU\s0 General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but \s-1WITHOUT ANY WARRANTY\s0; without even the implied warranty of
\&\s-1MERCHANTABILITY\s0 or \s-1FITNESS FOR A PARTICULAR PURPOSE.\s0  See the
\&\s-1GNU\s0 General Public License for more details.
.PP
You should have received a copy of the \s-1GNU\s0 General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.
.SH "Author"
.IX Header "Author"
T.v.Dein <tom \s-1AT\s0 vondein \s-1DOT\s0 org>


kleingebaeck.go
package main

var manpage = `
NAME
    kleingebaeck - kleinanzeigen.de backup tool

SYNOPSYS
        Usage: kleingebaeck [-dvVhmoc] [<ad-listing-url>,...]
        Options:
        -u --user    <uid>      Backup ads from user with uid <uid>.
        -d --debug              Enable debug output.
        -v --verbose            Enable verbose output.
        -o --outdir  <dir>      Set output dir (default: current directory)
        -l --limit   <num>      Limit the ads to download to <num>, default: load all.
        -c --config  <file>     Use config file <file> (default: ~/.kleingebaeck).
           --ignoreerrors       Ignore HTTP errors, may lead to incomplete ad backup.
        -f --force              Overwrite images and ads even if the already exist.
        -m --manual             Show manual.
        -h --help               Show usage.
        -V --version            Show program version.

DESCRIPTION
    This tool can be used to backup ads on the german ad page
    <https://kleinanzeigen.de>.

    It downloads all (or only the specified ones) ads of one user into a
    directory, each ad into its own subdirectory. The backup will contain a
    textfile Adlisting.txt which contains the ad contents such as title,
    body, price etc. All images will be downloaded as well.

CONFIGURATION
    You can create a config file to save typing. By default
    "~/.kleingebaeck" is being used but you can specify one with "-c" as
    well. We use TOML as our configuration language. See
    <https://toml.io/en/>.

    Format is pretty simple:

        user = 1010101
        loglevel = verbose
        outdir = "test"
        useragent = "Mozilla/5.0"
        template = """
        Title: {{.Title}}
        Price: {{.Price}}
        Id: {{.ID}}
        Category: {{.Category}}
        Condition: {{.Condition}}
        Created: {{.Created}}

        {{.Text}}
        """

    Be careful if you want to change the template. The variable is a
    multiline string surrounded by three double quotes. You can left out
    certain fields and use any formatting you like. Refer to
    <https://pkg.go.dev/text/template> for details how to write a template.
    Also read the TEMPLATES section below.

    If you're on windows and want to customize the output directory, put it
    into single quotes to avoid the backslashes interpreted as escape chars
    like this:

        outdir = 'C:\Data\Ads'

TEMPLATES
    Various parts of the configuration can be modified using templates: the
    output directory, the ad directory and the ad listing itself.

  OUTPUT DIR TEMPLATE
    The config varialbe "outdir" or the command line parameter "-o" take a
    template which may contain:

    "{{.Year}}"
    "{{.Month}}"
    "{{.Day}}"

    That way you can create a new output directory for every backup run. For
    example:

        outdir = "/home/backups/ads-{{.Year}}-{{.Month}}-{{.Day}}"

    Or using the command line flag:

        -o "/home/backups/ads-{{.Year}}-{{.Month}}-{{.Day}}"

    The default value is "." - the current directory.

  AD DIRECTORY TEMPLATE
    The ad directory name can be modified using the following ad values:

    {{.Price}}
    {{.ID}}
    {{.Category}}
    {{.Condition}}
    {{.Created}}
    {{.Slug}}
    {{.Text}}

    It can only be configured in the config file. By default only
    "{{.Slug}}" is being used, this is the title of the ad in url format.

  AD NAME TEMPLATE
    The name of the directory per ad can be tuned as well:

    "{{.Year}}"
    "{{.Month}}"
    "{{.Day}}"
    "{{.Slug}}"
    "{{.Category}}"
    "{{.ID}}"

  AD TEMPLATE
    The ad listing itself can be modified as well, using the same variables
    as the ad name template above.

    This is the default template:

        Title: {{.Title}}
        Price: {{.Price}}
        Id: {{.ID}}
        Category: {{.Category}}
        Condition: {{.Condition}}
        Type: {{.Type}}
        Created: {{.Created}}
        Expire: {{.Expire}}
    
        {{.Text}}

    The config parameter to modify is "template". See example.conf in the
    source repository. Please take care, since this is a multiline string.
    This is how it shall look if you modify it:

        template="""
        Title: {{.Title}}
    
        {{.Text}}
        """

    That is, the content between the two """ chars is the template.

SETUP
    To setup the tool, you need to lookup your userid on kleinanzeigen.de.
    Go to your ad overview page while NOT being logged in:

        https://www.kleinanzeigen.de/s-bestandsliste.html?userId=XXXXXX

    The XXXXX part is your userid.

    Put it into the configfile as outlined above. Also specify an output
    directory. Then just execute "kleingebaeck".

    You can use the -v option to get verbose output or -d to enable
    debugging.

ENVIRONMENT VARIABLES
    The following environment variables are considered:

        KLEINGEBAECK_USER
        KLEINGEBAECK_DEBUG
        KLEINGEBAECK_VERBOSE
        KLEINGEBAECK_OUTDIR
        KLEINGEBAECK_LIMIT
        KLEINGEBAECK_CONFIG
        KLEINGEBAECK_IGNOREERRORS

    Please note, that they take precedence over config file, but commandline
    flags take precedence over env!

BUGS
    In order to report a bug, unexpected behavior, feature requests or to
    submit a patch, please open an issue on github:
    <https://codeberg.org/scip/kleingebaeck/issues>.

    Please repeat the failing command with debugging enabled "-d" and
    include the output in the issue.

LIMITATIONS
    The "kleingebaeck" doesn't currently check if it has downloaded a file
    already, so it downloads everything again every time you execute it. Be
    aware of it. This will change in the future.

    Also there's currently no parallelization implemented. This will change
    in the future.

LICENSE
    Copyright 2023-2025 Thomas von Dein

    This program is free software: you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation, either version 3 of the License, or (at your
    option) any later version.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
    Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program. If not, see <http://www.gnu.org/licenses/>.

Author
    T.v.Dein <tom AT vondein DOT org>

`
